File: 0129-sum-root-to-leaf-numbers/0129-sum-root-to-leaf-numbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int sumNumbers(TreeNode* root) {
        int total = 0;  
        string curr = "";
        helper(root,total,curr);
        return total; 
    }

    void helper(TreeNode* root, int& total, string& curr){
        if(!root) return; 

        curr += to_string(root->val);

        helper(root->left, total, curr); 
        helper(root->right, total, curr); 

        if(!root->left && !root->right){
            //cout << curr << endl; 
            total += stoi(curr); 
        }
        curr.pop_back(); 
    }
};

File: 0129-sum-root-to-leaf-numbers/0129-sum-root-to-leaf-numbers.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int answer = 0; 
    public int sumNumbers(TreeNode root) {
        if(root.left == null && root.right == null) return root.val; 
        sumNumbersHelper(root.left, root.val); 
        sumNumbersHelper(root.right, root.val); 
        return answer; 
    }

    public void sumNumbersHelper(TreeNode root, int val){
        if(root == null) return; 

        val = val * 10 + (root.val); 
        sumNumbersHelper(root.left,val); 
        sumNumbersHelper(root.right,val);

        if(root.left == null && root.right == null){
            answer += val; 
        }
    }
}